File: backend/internal/shared/domain/plate_test.go
package domain_test

import (
	"testing"

	"fleetmonitoring/backend/internal/shared/domain"
)

// Covers [SPEC-001: AC-004, FR-004, BR-001]
// TEST-004 — Plate validation: regex ^[A-Z]{3}[0-9]{3}$, normalized uppercase.
// Valid: GTP890, TTY423. Invalid: GTP89, 123, empty, wrong length, symbols.

func TestPlate(t *testing.T) {
	t.Run("accepts valid plates", func(t *testing.T) {
		// Covers [SPEC-001: AC-004, FR-001, BR-001]
		valid := []string{"GTP890", "TTY423", "ABC123", "ZZZ999", "AAA000"}
		for _, input := range valid {
			input := input
			t.Run(input, func(t *testing.T) {
				// Arrange
				// Act
				plate, err := domain.ParsePlate(input)

				// Assert
				if err != nil {
					t.Fatalf("expected valid plate %q to parse without error, got %v", input, err)
				}
				if string(plate) != input {
					t.Fatalf("expected plate %q got %q", input, string(plate))
				}
				if !plate.IsValid() {
					t.Fatalf("expected IsValid true for %q", input)
				}
			})
		}
	})

	t.Run("normalizes lowercase to uppercase", func(t *testing.T) {
		// Covers [SPEC-001: AC-004, BR-001]
		cases := []struct {
			input string
			want  string
		}{
			{"gtp890", "GTP890"},
			{"tty423", "TTY423"},
			{"abc123", "ABC123"},
			{"Gtp890", "GTP890"},
			{"gTp890", "GTP890"},
		}
		for _, tc := range cases {
			tc := tc
			t.Run(tc.input, func(t *testing.T) {
				// Arrange
				// Act
				plate, err := domain.ParsePlate(tc.input)

				// Assert
				if err != nil {
					t.Fatalf("expected %q to normalize and be valid, got error %v", tc.input, err)
				}
				if string(plate) != tc.want {
					t.Fatalf("expected normalized %q got %q", tc.want, string(plate))
				}
			})
		}
	})

	t.Run("normalizes and validates trimmed input", func(t *testing.T) {
		// Covers [SPEC-001: AC-004, BR-001]
		// Arrange
		input := " gtp890 "

		// Act
		plate, err := domain.ParsePlate(input)

		// Assert
		if err == nil {
			t.Fatalf("expected error for plate with spaces %q, got plate %q", input, plate)
		}
	})

	t.Run("rejects invalid plates", func(t *testing.T) {
		// Covers [SPEC-001: AC-004, FR-004, BR-001]
		invalid := []struct {
			name  string
			input string
		}{
			{"too short GTP89", "GTP89"},
			{"numeric only 123", "123"},
			{"too long GTP8901", "GTP8901"},
			{"too long 4 letters", "ABCD123"},
			{"too short 2 letters", "AB1234"},
			{"lower invalid length", "ab123"},
			{"with dash", "GT-890"},
			{"with space", "GTP 890"},
			{"empty", ""},
			{"whitespace", "   "},
			{"letters only", "ABCDEF"},
			{"digits only 6", "123456"},
			{"mix wrong pattern ABC12", "ABC12"},
			{"mix wrong pattern 123ABC", "123ABC"},
			{"special char", "GTP89@ "},
			{"lower too short gtp89", "gtp89"},
		}
		for _, tc := range invalid {
			tc := tc
			t.Run(tc.name, func(t *testing.T) {
				// Arrange
				// Act
				plate, err := domain.ParsePlate(tc.input)

				// Assert
				if err == nil {
					t.Fatalf("expected error for invalid plate %q, got plate %q", tc.input, plate)
				}
			})
		}
	})

	t.Run("String returns normalized form", func(t *testing.T) {
		// Covers [SPEC-001: BR-001]
		// Arrange
		input := "gtp890"

		// Act
		plate, err := domain.ParsePlate(input)

		// Assert
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if plate.String() != "GTP890" {
			t.Fatalf("expected String() GTP890 got %q", plate.String())
		}
	})

	t.Run("IsValid matches ParsePlate validity", func(t *testing.T) {
		// Covers [SPEC-001: AC-004, BR-001]
		// Arrange
		validPlate := domain.Plate("GTP890")
		invalidPlate := domain.Plate("GTP89")

		// Act
		valid := validPlate.IsValid()
		invalid := invalidPlate.IsValid()

		// Assert
		if !valid {
			t.Fatalf("expected IsValid true for GTP890")
		}
		if invalid {
			t.Fatalf("expected IsValid false for GTP89")
		}
	})
}

File: mobile/src/lib/plate.test.ts
import { normalizePlate, isValidPlate } from './plate';

// Covers [SPEC-005: AC-001] - FR-001 BR-001/002 TS-001: Plate SSOT normalize + validate
describe('plate lib', () => {
  describe('normalizePlate', () => {
    it('normalizes lower case to upper case', () => {
      // Arrange
      const input = 'acf356';
      const expected = 'ACF356';

      // Act
      const result = normalizePlate(input);

      // Assert
      expect(result).toBe(expected);
    });

    it('trims whitespace and uppercases', () => {
      // Arrange
      const input = '  acf356 ';
      const expected = 'ACF356';

      // Act
      const result = normalizePlate(input);

      // Assert
      expect(result).toBe(expected);
    });

    it('trims tabs and mixed case', () => {
      // Arrange
      const input = '\t AcF356 \n';
      const expected = 'ACF356';

      // Act
      const result = normalizePlate(input);

      // Assert
      expect(result).toBe(expected);
    });
  });

  describe('isValidPlate', () => {
    it('returns true for valid plate ACF356', () => {
      // Arrange
      const plate = 'ACF356';

      // Act
      const result = isValidPlate(plate);

      // Assert
      expect(result).toBe(true);
    });

    it('returns true for valid plate after normalize lower case', () => {
      // Arrange
      const plate = 'acf356';

      // Act
      const result = isValidPlate(plate);

      // Assert
      expect(result).toBe(true);
    });

    it('returns false for too short plate ACF35 (5 chars)', () => {
      // Arrange
      const plate = 'ACF35';

      // Act
      const result = isValidPlate(plate);

      // Assert
      expect(result).toBe(false);
    });

    it('returns false for too long plate ACF3567 (7 chars)', () => {
      // Arrange
      const plate = 'ACF3567';

      // Act
      const result = isValidPlate(plate);

      // Assert
      expect(result).toBe(false);
    });

    it('returns false for invalid pattern 12ABC3', () => {
      // Arrange
      const plate = '12ABC3';

      // Act
      const result = isValidPlate(plate);

      // Assert
      expect(result).toBe(false);
    });

    it('returns false for empty string', () => {
      // Arrange
      const plate = '';

      // Act
      const result = isValidPlate(plate);

      // Assert
      expect(result).toBe(false);
    });

    it('returns false for plate with hyphen ACF-356', () => {
      // Arrange
      const plate = 'ACF-35';

      // Act
      const result = isValidPlate(plate);

      // Assert
      expect(result).toBe(false);
    });

    it('returns false for plate with space ACF 356', () => {
      // Arrange
      const plate = 'ACF 356';

      // Act
      const result = isValidPlate(plate);

      // Assert
      expect(result).toBe(false);
    });
  });
});
